Feature/new 188746 user settings

PhonicCanine opened the pull request:
Added more configurable settings, with defaults that mimic the default behavior of the Element webapp. In order to expose features, the config file in the web app repo is necessary.
To merge into the upstream repo, some documentation changes will be necessary.

**User Stories Completed:**
1. 188496: Web: As an Authenticated Practitioner, I should not be able to view certain Settings that are not applicable or do not need to be displayed to the user, so that I only see applicable settings2. 188488 - Web: As an Authenticated Practitioner, I should be able to personalise the General Settings, so that I can control certain system behaviours that will maximise my user experience
2. 188622: Web: As an Authenticated Practitioner, I should be able to personalise the Voice & Video Settings, so that I can control certain system behaviours that will maximise my user experience

**Work in Progress**
3. 188497 - Web: As an Authenticated Practitioner, I should be able to personalise the Notification Settings, so that I can control certain system behaviours that will maximise my user experience
4. 188663 - Web: As an Authenticated Practitioner, I should be able to personalise the Security & Privacy Settings, so that I can control certain system behaviours that will maximise my user experience
5. 188664 - Web: As an Authenticated Practitioner, I should be able to personalise the Help & About Settings, so that I can control certain system behaviours that will maximise my user experience

PhonicCanine committed:
Configurable User Settings completed for new version

PhonicCanine committed:
Merge branch 'master' of https://github.com/matrix-org/matrix-react-sdk into feature/new-188746-user-settings

jrschremmer commented on src/components/structures/UserMenu.tsx:
```diff
@@ -478,17 +490,7 @@ export default class UserMenu extends React.Component<IProps, IState> {
-                        width={16}
-                    />
-                </AccessibleTooltipButton>
+                {SettingsStore.getValue(UIFeature.ThemeChanging) && themebutton}
```
In wanting to minimise changes to the code base, does the button definition need to be encapsulated into a separate variable or could it just be the following, as the themebutton is not referenced else where:
```
{SettingsStore.getValue(UIFeature.ThemeChanging) && <AccessibleTooltipButton
            className="mx_UserMenu_contextMenu_themeButton"
            onClick={this.onSwitchThemeClick}
            title={this.state.isDarkTheme ? _t("Switch to light mode") : _t("Switch to dark mode")}
        >
            <img
                src={require("../../../res/img/element-icons/roomlist/dark-light-mode.svg")}
                alt={_t("Switch theme")}
                width={16}
            />
        </AccessibleTooltipButton>
```

jrschremmer replied:
I have implemented this suggestion, please review and suggest any changes.

sppanday replied:
I had brief discussion few days ago. Usually with template definitions , variable definition is used so that it makes code more readable in if statement but We get choice of two: One is to make code readable and make that condition reusable for future but at the same time other decision would be to stick to matrix pattern to avoid merge conflict. I would also incline to your approach Jasen just to avoid Merge Conflict but I had to write everything from scratch which doesnt apply to this case, I would define html as variable and inject when needed. Somewhere around other javascript files I have found matrix using variable definition , so there has not been consistency in their approach.

jrschremmer replied:
Surendra I agree with the variable approach if we were starting from scratch ... you are always going to get inconsistencies with multiple developers, and more importantly multiple code reviewers

jrschremmer commented on src/settings/UIFeature.ts:
```diff
@@ -32,4 +32,25 @@ export enum UIFeature {
     Communities = "UIFeature.communities",
     AdvancedSettings = "UIFeature.advancedSettings",
     RoomHistorySettings = "UIFeature.roomHistorySettings",
+    ThemeChanging = "UIFeature.themeChanging",
```
To be consistent with the ChangeDisplayName variable name, I suggest ThemeChanging should be renamed to ChangeTheme

jrschremmer replied:
I have implemented this suggestion, please review and suggest any changes.

sppanday replied:
Hi Jasen, I agree with your variable naming convention.

jrschremmer commented on src/settings/UIFeature.ts:
```diff
@@ -32,4 +32,25 @@ export enum UIFeature {
     AdvancedSettings = "UIFeature.advancedSettings",
     RoomHistorySettings = "UIFeature.roomHistorySettings",
+    ThemeChanging = "UIFeature.themeChanging",
+    FontChanging = "UIFeature.fontChanging",
```
To be consistent with the ChangeDisplayName variable name, I suggest FontChanging should be renamed to ChangeFont

jrschremmer replied:
I have implemented this suggestion, please review and suggest any changes.

sppanday replied:
Hi Jasen, I agree with your variable naming convention.

jrschremmer commented on src/settings/watchers/ThemeWatcher.ts:
```diff
@@ -103,7 +103,7 @@ export default class ThemeWatcher {
         // to ensure the ordering otherwise works.
         const systemThemeExplicit = SettingsStore.getValueAt(
-            SettingLevel.DEVICE, "use_system_theme", null, false, true);
+            SettingLevel.DEVICE, "use_system_theme", null);
```
To minimise changes, the default value of explicit which is false, can be kept and the value of excludeDefault could be SettingsStore.getValue(UIFeature.ChangeTheme)

jrschremmer replied:
I have implemented this suggestion, please review and suggest any changes.

PhonicCanine replied:
I would probably prefer extract out another user setting (themeFallsOnConfig or something, and use `SettingsStore.getValue(UIFeature.themeFallsOnConfig) || SettingsStore.getValue(UIFeature.ChangeTheme)`) because I think there's a use-case for setting a default theme and then allowing the user to override it, that is distinct and different from the case of when the theme is set and the user can't change it.

The original change allowed this use case by default; but ultimately I think it's a relatively unimportant point, so I'm happy with the suggested code change.

sppanday replied:
Hi Jasen, I had same suggestion in our chat for Joseph to leave default changes and add extra if needed than removing booleans. I am glad you have chosen that approach as well.

sppanday replied:
@PhonicCanine , yeah, when we add new things we consider "minimise changes", "minimise code merge conflict", then think of other aspects they dont negatively impact these two.

jrschremmer commented on src/settings/watchers/ThemeWatcher.ts:
```diff
@@ -114,7 +114,7 @@ export default class ThemeWatcher {
         // level for the setting to ensure the ordering otherwise works.
         const themeExplicit = SettingsStore.getValueAt(
-            SettingLevel.DEVICE, "theme", null, false, true);
+            SettingLevel.DEVICE, "theme", null);
```
To minimise changes, the default value of explicit which is false, can be kept and the value of excludeDefault could be SettingsStore.getValue(UIFeature.ChangeTheme)

jrschremmer replied:
I have implemented this suggestion, please review and suggest any changes.

PhonicCanine replied:
Same as above comment. This solution works for me, but I would prefer also cater to the use case that allows a default theme to be specified, and allows the user to change it.

sppanday replied:
same as previous comment above

jrschremmer commented on src/settings/UIFeature.ts:
```diff
@@ -32,4 +32,25 @@ export enum UIFeature {
+    PreferencesPaneHiddenFlags = "UIFeature.preferencesPaneHiddenFlags",
+    ShowAdvancedCallSettings = "UIFeature.showAdvancedCallSettings",
+    SecurityAllowSessionEdits = "UIFeature.securityAllowSessionEdits",
+    EnableEncryptionSettings = "UIFeature.enableEncryptionSettings",
```
For consistency I suggest we rename EnableEncryptionSettings to ShowEncryptionSettings

jrschremmer replied:
I have implemented this suggestion, please review and suggest any changes.

jrschremmer commented on src/components/views/settings/account/EmailAddresses.js:
```diff
@@ -83,6 +85,13 @@ export class ExistingEmailAddress extends React.Component {
     };
 
     render() {
+        if (!SettingsStore.getValue(UIFeature.EditEmailAddresses)) {
```
To minimise changes and centralise code, I'd suggest merging this block of code with line 114

jrschremmer replied:
I have implemented this suggestion, please review and suggest any changes.

jrschremmer commented on src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:
```diff
@@ -93,9 +93,9 @@ export default class AppearanceUserSettingsTab extends React.Component<IProps, I
         const themeChoice: string = SettingsStore.getValue("theme");
         const systemThemeExplicit: boolean = SettingsStore.getValueAt(
-            SettingLevel.DEVICE, "use_system_theme", null, false, true);
+            SettingLevel.DEVICE, "use_system_theme");
```
Same as the suggested updates to ThemeWatcher

jrschremmer replied:
I have implemented this suggestion, please review and suggest any changes.

jrschremmer commented on src/components/views/settings/tabs/user/GeneralUserSettingsTab.js:
```diff
@@ -283,17 +283,20 @@ export default class GeneralUserSettingsTab extends React.Component {
-                <p className="mx_SettingsTab_subsectionText">
-                    {passwordChangeText}
-                </p>
-                {passwordChangeForm}
```
I suggest you change this and use the wrapping pattern you have used elsewhere e.g. <> to avoid repeating SettingsStore.getValue(UIFeature.ShowPasswordReset) and minimise changes

jrschremmer replied:
I have implemented this suggestion, please review and suggest any changes.

jrschremmer commented on src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:
```diff
@@ -405,12 +407,12 @@ export default class AppearanceUserSettingsTab extends React.Component<IProps, I
             );
             advanced = <>
-                <SettingsFlag
+                {(!UIFeature.PullUpCompactLayout) && <SettingsFlag
```
It appears this is missing SettingsStore.getValue(

jrschremmer replied:
I have implemented this suggestion, please review and suggest any changes.

jrschremmer commented on src/components/views/settings/account/PhoneNumbers.js:
```diff
@@ -260,20 +264,22 @@ export default class PhoneNumbers extends React.Component {
+                            />
+                        </div>
+                    </form>
+                }
```
I suggest you change this and use the wrapping pattern you have used elsewhere e.g. <> to avoid repeating SettingsStore.getValue(UIFeature.EditPhoneNumbers) and minimise changes

jrschremmer replied:
I have implemented this suggestion, please review and suggest any changes.

jrschremmer committed:
Code Review: See https://github.com/fhirfactory/pegacorn-communicate-matrix-react-sdk/pull/4
Minimise changes and put our changes on new lines whenever possible to assist with automatic merging;
Standardise settings names and how the code is surrounded with the lookup of the new settings that turn the behaviour on or off;

jrschremmer commented on src/components/views/settings/tabs/user/VoiceUserSettingsTab.js:
```diff
@@ -186,10 +186,13 @@ export default class VoiceUserSettingsTab extends React.Component {
-                           value={this.state.activeAudioInput || defaultDevice}
-                           onChange={this._setAudioInput}>
-                        {this._renderDeviceOptions(audioInputs, 'audioInput')}
+                    <Field
```
I can't see a functional change in these lines.  To minimise merge issues with the base repository, we need to minimise changes.  We you please revert the changes to the Field tag, unless there is a functional change that I have missed.

sppanday replied:
Done

jrschremmer commented on src/components/views/settings/tabs/user/VoiceUserSettingsTab.js:
```diff
@@ -215,19 +218,34 @@ export default class VoiceUserSettingsTab extends React.Component {
-                            onChange={this._changeFallbackICEServerAllowed}
-                        />
-                    </>}
+                    {SettingsStore.getValue(
```
Please minimise any changes to existing code.  Is you IDE set to wrap lines with a short line length?  Our standard for line length wrapping is 160 characters.
Is there a difference between SettingsStore.getValue(UIFeature.ShowAdvancedCallSettings) && <>
and SettingsStore.getValue(UIFeature.ShowAdvancedCallSettings) ? ... : null ?  As the bool && <> is used throughout the existing code.

sppanday replied:
&& would provide conditional way of handling jsx, and :null would be similar too but yeah tried hard to stick to matrix's way but did not work, now its working so made changes to <> too as well. This is react fragment syntax which is specific for certain purpose, div can be used as well . You are right in suggesting to use </> as matrix uses that too but sometimes situation makes usage of div unavoidable.

jrschremmer commented on src/settings/Settings.ts:
```diff
@@ -817,4 +817,12 @@ export const SETTINGS: {[setting: string]: ISetting} = {
     },
+    [UIFeature.ShowOneToOneRTCCall]: {
+        supportedLevels: LEVELS_UI_FEATURE,
+        default: false,
```
Please set the defaults to the element base repo behaviour, which I believe would be true instead of false

sppanday replied:
Done

jrschremmer commented on src/components/views/settings/tabs/user/VoiceUserSettingsTab.js:
```diff
@@ -215,19 +218,34 @@ export default class VoiceUserSettingsTab extends React.Component {
+                    ) ? (
+                        <div>
+                            {SettingsStore.getValue(
+                                UIFeature.ShowLocalVideoFeed
```
For clarify, please rename ShowLocalVideoFeed to ShowSettingMirrorLocalVideoFeed

jrschremmer commented on src/components/views/settings/tabs/user/VoiceUserSettingsTab.js:
```diff
@@ -215,19 +218,34 @@ export default class VoiceUserSettingsTab extends React.Component {
+                                />
+                            )}
+                            {SettingsStore.getValue(
+                                UIFeature.ShowOneToOneRTCCall
```
For clarify, please rename ShowOneToOneRTCCall to ShowSettingAllowPeerToPeerForOneToOneCalls

sppanday committed:
added config to hide 1-1 call,
 local video feed 188662 in audio and video settings

sppanday committed:
added feature  documentation for communicate web

sppanday committed:
variable naming convention changes for user story 188662

sppanday committed:
changed UIFeature variables to camelCase

changed variable name to camelCase to follow existing pattern

sppanday committed:
updated documentation for 188662

jrschremmer committed:
Minimise changes to existing code in master;
Change the defaults of the new toggles to be true to match what the current master code does;
Standardise the case of the UIFeature configuration names
